File: src/main/java/tmi/recipe/parser/GenericCrafterParser.java
package tmi.recipe.parser;

import arc.struct.Seq;
import mindustry.type.ItemStack;
import mindustry.type.LiquidStack;
import mindustry.world.Block;
import mindustry.world.blocks.production.GenericCrafter;
import tmi.recipe.Recipe;
import tmi.recipe.RecipeType;

public class GenericCrafterParser extends ConsumerParser<GenericCrafter>{
  @Override
  public boolean isTarget(Block content) {
    return content instanceof GenericCrafter;
  }

  @Override
  public Seq<Recipe> parse(GenericCrafter crafter) {
    Recipe res = new Recipe(RecipeType.factory)
        .setBlock(getWrap(crafter))
        .setTime(crafter.craftTime);

    registerCons(res, crafter.consumers);

    if (crafter.outputItems == null) {
      if (crafter.outputItem != null) res.addProduction(getWrap(crafter.outputItem.item), crafter.outputItem.amount);
    }
    else {
      for (ItemStack item : crafter.outputItems) {
        res.addProduction(getWrap(item.item), item.amount);
      }
    }

    if (crafter.outputLiquids == null) {
      if (crafter.outputLiquid != null) res.addProductionPresec(getWrap(crafter.outputLiquid.liquid), crafter.outputLiquid.amount);
    }
    else {
      for (LiquidStack liquid : crafter.outputLiquids) {
        res.addProductionPresec(getWrap(liquid.liquid), liquid.amount);
      }
    }

    return Seq.with(res);
  }
}

File: src/main/kotlin/tmi/recipe/parser/GenericCrafterParser.kt
package tmi.recipe.parser

import arc.struct.Seq
import mindustry.world.Block
import mindustry.world.blocks.production.GenericCrafter
import tmi.recipe.Recipe
import tmi.recipe.RecipeType

open class GenericCrafterParser : ConsumerParser<GenericCrafter>() {
  override fun isTarget(content: Block): Boolean {
    return content is GenericCrafter
  }

  override fun parse(content: GenericCrafter): Seq<Recipe> {
    val res = Recipe(
      recipeType = RecipeType.factory,
      ownerBlock = content.getWrap(),
      craftTime = content.craftTime
    )

    registerCons(res, *content.consumers)

    if (content.outputItems == null) {
      if (content.outputItem != null) res.addProductionInteger(content.outputItem.item.getWrap(), content.outputItem.amount)
    }
    else {
      for (item in content.outputItems) {
        res.addProductionInteger(item.item.getWrap(), item.amount)
      }
    }

    if (content.outputLiquids == null) {
      if (content.outputLiquid != null) res.addProductionPersec(
        content.outputLiquid.liquid.getWrap(),
        content.outputLiquid.amount
      )
    }
    else {
      for (liquid in content.outputLiquids) {
        res.addProductionPersec(liquid.liquid.getWrap(), liquid.amount)
      }
    }

    return Seq.with(res)
  }
}
